chapters/VK_EXT_present_timing/PresentTimeInfo.txt

```diff
@@ -0,0 +1,184 @@
+slink:VkAbsolutePresentTimeEXT or a slink:VkRelativePresentTimeEXT
+structure, defined as:
+
+include::{generated}/api/structs/VkPresentTimeEXT.txt[]
+
+include::{generated}/validity/structs/VkPresentTimeEXT.txt[]
+--
+
+[open,refpage='VkAbsolutePresentTimeEXT',desc='Specifying the earliest time an image should be presented',type='structs']
+--
+
+The sname:VkAbsolutePresentTimeEXT structure is defined as:
+
+include::{generated}/api/structs/VkAbsolutePresentTimeEXT.txt[]
+
+  * pname:presentID is an application-provided identification value.
+    If non-zero, flink:vkGetPastPresentationTimingEXT will return a
+    slink:VkPastPresentationTimingEXT struct with the timing information
+    associated with the present to the associated swapchain for this
+    call to flink:vkQueuePresentKHR, including this value being in the
+    pname:presentID member of the slink:VkPastPresentationTimingEXT
+    struct.
+    In order to be useful to the application, it should: be unique within
+    some period of time that is meaningful to the application.
+  * pname:targetPresentTime, if non-zero, specifies the earliest time the
+    application wants the image to be displayed to the user.
+    pname:targetPresentTime is a time in nanoseconds, according to the
+    time-domain being used.
+    A value of zero specifies that the presentation engine may: display the
+    image at any time.
+  * pname:idealPresentTime provides an indication to the presentation
+    engine of what the application desires its IPD to become.
+    An application can: set pname:idealPresentTime to the same value
+    as pname:targetPresentTime, indicating that the application
+    desires its IPD to remain steady.
+    A value earlier than
+    pname:targetPresentTime indicates that the application desires to
+    shorten its IPD.
+    A value later than pname:targetPresentTime
+    indicates that the application desires to lengthen its IPD.
+    The presentation engine will provide feedback on this value in the
+    data returned by flink:vkGetPastPresentationTimingEXT.
+  * pname:presentSlop is the period of time in nanoseconds before
+    pname:targetPresentTime that the presentation engine may: display
+    the image.
+
+If pname:targetPresentTime is non-zero, the presentation engine should:
+not display the image to the user at a time earlier than
+(pname:targetPresentTime - pname:presentSlop).
+
+Setting pname:targetPresentTime to zero is useful when the application
+desires to provide a non-zero pname:presentID in order to get back timing
+information from flink:vkGetPastPresentationTimingEXT, but does not wish
+to set a pname:targetPresentTime.
+
+Providing a pname:presentSlop enables the application to indicate to the
+presentation engine to display the image within this period if a vertical
+blanking period occurs on FRR displays.
+
+[NOTE]
+.Note
+====
+The pname:presentSlop is used to avoid unintentionally missing a vertical
```

Comment:
Is there a reason why the presentation engine needs both `targetPresentTime` and `presentSlop`? Why can't the application just provide a target of `targetPresentTime - presentSlop`?

I was wondering if the reason was VRR, but there's already `idealPresentTime` that could be used in this case.

Reply:
I asked a similar question on the discussion for the new wayland present-timing protocol which i think is supposed to provide the foundation for VK_EXT_present_timing for Wayland/WSI.

My understanding would be that presentSlop is only meant for FRR displays, to avoid skipped frames if targetPresentTime happens to lie inside or just after the end of a vblank, due to numerical inaccuracies, or slightly noisy presentation feedback timestamps used as basis for calculating a new targetPresentTime. E.g., given that actualPresentTime reported by the driver would mark the end of a vblank/start of a new active scanout period, doing an animation that sets targetPresentTime = actualPresentTime + some_multiple_of_a_refresh_duration could easily miss the present for the wanted target vblank if actualPresentTime is a few microseconds too late or some_multiple_of_a_refresh_duration is slightly too large due to accumulated numerical roundoff error.

So my understanding of presentSlop for a FRR display would be that iff the time interval [targetPresentTime - presentSlop ; targetPresentTime] intersects with a vblank, then the presentation engine should target that intersected vblank for presenting. But if the interval does not intersect with a vblank, then the semantic should be that a present must not happen before targetPresentTime. For a sufficiently large presentSlop, e.g., 1 msec, that would avoid accidentally missing a target vblank and thereby a skipped frame, while still preserving which vblank the application intended to target for a present.

E.g., on Linux with the open-source display stack, a display server could compare vblank timestamps against such an interval and in the case of intersection it would schedule a page-flip for the refresh interval anywhere *before* the start of the intersecting vblank interval and leave it to the kms drivers and gpu to wait for completion of all needed fences and lock to the right vsync to trigger the page-flip in hardware.

Right understanding? Or a misunderstanding of the use case for presentSlop?

My own software implements a similar mechanism for FRR mode by advising to use this formula to calculate a targetPresentTime, based on actualPresentTime from a previous present, if it wants to present every 'waitFrames' video refresh cycles:

targetPresentTime = actualPresentTime + waitFrames * refreshDuration - 0.5 * refreshDuration

presentSlop would be the equivalent of the shift of "-0.5 * refreshDuration", but more elegant and robust at the driver level and wrt. interfacing with the application. Assuming i understand the intended use of presentSlop?